src/parser.rs

```diff
@@ -195,16 +204,26 @@ fn parse_predicate(pair: Pair<Rule>) -> ParseResult<Predicate> {
         lhs,
         rhs: if op == BinaryOperator::Regex {
             if let Value::String(s) = rhs {
-                let r = Regex::new(&s).map_err(|e| {
-                    ParseError::new_from_span(
-                        ErrorVariant::CustomError {
-                            message: e.to_string(),
-                        },
-                        rhs_pair.as_span(),
-                    )
-                })?;
+                let regex_rc = match regex_cache.get(&s) {
+                    Some(stored_regex_rc) => stored_regex_rc.clone(),
+                    _ => {
+                        let r = Regex::new(&s).map_err(|e| {
+                            ParseError::new_from_span(
+                                ErrorVariant::CustomError {
+                                    message: e.to_string(),
+                                },
+                                rhs_pair.as_span(),
+                            )
+                        })?;
+
+                        let rc = Rc::new(r);
 
-                Value::Regex(r)
+                        regex_cache.insert(s, rc.clone());
+                        rc
+                    }
+                };
+
+                Value::Regex(regex_rc)
```

Comment:
We could use https://doc.rust-lang.org/stable/std/collections/struct.HashMap.html#method.entry to simply the logic.

Reply:
Actually... I don't think we can :/ 
I've wanted to use `entry` but it consumes string. I'd have to clone it but then it defeats the purpose of this optimization.
There's this big thread to that tried to address this by COW and lookup by reference but I don't think it got anywhere: https://github.com/rust-lang/rfcs/pull/1769

Reply:
@nowNick Ah, you are right, the `entry()` consumes the key.